multi_imbalance/ensemble/ovo.py

```diff
@@ -0,0 +1,176 @@
+                X_filtered, y_filtered = X[filtered_indices], y[filtered_indices]
+                if self.should_perform_ovesampling(first_class, second_class):
+                    X_filtered, y_filtered = self._oversample(X_filtered, y_filtered, strategy=self.oversample_binary)
+                self._binary_classifiers[row][col].fit(X_filtered, y_filtered)
+
+    def _get_classifier(self):
+        allowed_classifiers = ('CART', 'NB', 'KNN')
+        if self.binary_classifier not in allowed_classifiers:
+            raise ValueError("Unknown binary classifier: %s, expected to be one of %s."
+                             % (self.binary_classifier, allowed_classifiers))
+        elif self.binary_classifier == 'CART':
+            decision_tree_classifier = DecisionTreeClassifier()
+            return decision_tree_classifier
+        elif self.binary_classifier == 'NB':
+            gnb = GaussianNB()
+            return gnb
+        elif self.binary_classifier == 'KNN':
+            knn = KNeighborsClassifier(n_neighbors=self.n_neighbors)
+            return knn
+
+    def _perform_voting(self, binary_outputs_matrix):
+        allowed_voting_strategies = ('max',)
+        if self.voting_strategy not in allowed_voting_strategies:
+            raise ValueError("Unknown voting strategy: %s, expected to be one of %s."
+                             % (self.voting_strategy, allowed_voting_strategies))
+        elif self.voting_strategy == 'max':
+            return self._perform_max_voting(binary_outputs_matrix)
+
+    def _perform_max_voting(self, binary_outputs_matrix):
+        scores = np.zeros(len(self._labels))
+        for clf_1 in range(len(binary_outputs_matrix)):
+            for clf_2 in range(clf_1):
+                scores[self._labels.tolist().index(binary_outputs_matrix[clf_1][clf_2])] += 1
+        return self._labels[np.argmax(scores)]
+
+    def _oversample(self, X, y, strategy=None):
+        allowed_oversampling = [None, 'globalCS', 'SMOTE']
+        if strategy not in allowed_oversampling:
+            raise ValueError("Unknown matrix generation encoding: %s, expected to be one of %s."
+                             % (strategy, allowed_oversampling))
+        elif strategy is None:
+            return X, y
+        elif strategy == 'globalCS':
+            gcs = GlobalCS()
+            return gcs.fit_transform(X, y)
+        elif strategy == 'SMOTE':
+            return self._smote_oversample_if_possible_random_otherwise(X, y)
+
+    def _smote_oversample_if_possible_random_otherwise(self, X, y):
+        if min(np.unique(y, return_counts=True)[1]) < 2:
+            return GlobalCS().fit_transform(X, y)
+
+        n_neighbors = 3 if min(np.unique(y, return_counts=True)[1]) > 3 else 1
+        smote = SMOTE(k_neighbors=n_neighbors)
+        smote.fit(X, y)
+        return smote.fit_resample(X, y)
+
+    def should_perform_ovesampling(self, first_class, second_class):
+        allowed_oversample_between = ['all', 'maj-min']
```

Comment:
If it is some meta information you can set is as a static variable in class

Reply:
thanks, done